# Empresa de relógios
## Alunos

| Nome                        | Matrícula |
|-----------------------------|-----------|
| Luiz Filipe Bartelega Penha | 202111082 |
| Vitor Pires Zini            | 202110169 |

In [ ]:
!pip install gurobipy

In [ ]:
from gurobipy import Model, GRB

modelo = Model("relogio")

In [ ]:

relogio_pedestal_var = modelo.addVar(name="relogio pedestal", vtype=GRB.CONTINUOUS, lb=0)
relogio_parede_var = modelo.addVar(name="relogio parede", vtype=GRB.CONTINUOUS, lb=0)

In [ ]:
modelo.setObjective(300 * relogio_pedestal_var + 200 * relogio_parede_var, GRB.MAXIMIZE)

In [ ]:
modelo.addConstr(6 * relogio_pedestal_var + 4 * relogio_parede_var <= 40, name="horas do João")

<gurobi.Constr *Awaiting Model Update*>

In [ ]:
modelo.addConstr(8 * relogio_pedestal_var + 4 * relogio_parede_var <= 40, name="horas da Ana")

<gurobi.Constr *Awaiting Model Update*>

In [ ]:
modelo.addConstr(3 * relogio_pedestal_var + 3 * relogio_parede_var <= 20, name="horas da Lydia")

<gurobi.Constr *Awaiting Model Update*>

In [ ]:
modelo.optimize()

Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 3 rows, 2 columns and 6 nonzeros
Model fingerprint: 0x6103ff60
Coefficient statistics:
  Matrix range     [3e+00, 8e+00]
  Objective range  [2e+02, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 4e+01]
Presolve time: 0.01s
Presolved: 3 rows, 2 columns, 6 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    5.0000000e+32   4.250000e+30   5.000000e+02      0s
       2    1.6666667e+03   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.666666667e+03


In [ ]:
if(modelo.status == GRB.OPTIMAL):
  print(f"Relogio pedestal: {relogio_pedestal_var.X}")
  print(f"Relogio parede: {relogio_parede_var.X}")
  print(f"Valor ótimo: {modelo.objVal}")


Relogio pedestal: 3.333333333333333
Relogio parede: 3.3333333333333335
Valor ótimo: 1666.6666666666665


# Questão 2


In [ ]:
print("\nIntervalos de Custos (Vetor de Custos):")
for variavel in modelo.getVars():
  print(f"Variável {variavel.VarName}:")
  print(f"  lower bound do custo (SAObjLow): {variavel.SAObjLow}")
  print(f"  upper bound do custo (SAObjUp): {variavel.SAObjUp}")

print("\nIntervalos de Recursos (Vetor de Recursos):")
for restricao in modelo.getConstrs():
  print(f"Restrição {restricao.ConstrName}:")
  print(f"  lower bound do recurso (SARHSLow): {restricao.SARHSLow}")
  print(f"  upper bound do recurso (SARHSUp): {restricao.SARHSUp}")
  print(f"  Variavel dual '{restricao.ConstrName}': Dual = {restricao.Pi}")


Intervalos de Custos (Vetor de Custos):
Variável relogio pedestal:
  lower bound do custo (SAObjLow): 200.0
  upper bound do custo (SAObjUp): 400.0
Variável relogio parede:
  lower bound do custo (SAObjLow): 150.0
  upper bound do custo (SAObjUp): 300.0

Intervalos de Recursos (Vetor de Recursos):
Restrição horas do João:
  lower bound do recurso (SARHSLow): 33.333333333333336
  upper bound do recurso (SARHSUp): inf
  Variavel dual 'horas do João': Dual = 0.0
Restrição horas da Ana:
  lower bound do recurso (SARHSLow): 26.666666666666668
  upper bound do recurso (SARHSUp): 53.33333333333333
  Variavel dual 'horas da Ana': Dual = 25.0
Restrição horas da Lydia:
  lower bound do recurso (SARHSLow): 15.0
  upper bound do recurso (SARHSUp): 30.0
  Variavel dual 'horas da Lydia': Dual = 33.333333333333336


## Explicação

Em relação aos custos, o lucro do relógio de pedestal apresenta uma margem de flexibilidade significativa, podendo variar entre R$200 e R$ 400 sem comprometer a solução ótima. De maneira similar, o lucro do relógio de parede permite oscilações entre R$ 150 e R$ 300. Caso os valores ultrapassem esses intervalos, outra solução ótima será produzida.

Quanto à disponibilidade de horas de trabalho, a análise revelou aspectos importantes para cada sócio:

João tem como limite mínimo 33,33 horas semanais e nenhum limite máximo de horas. Cada hora adicional que ele trabalha contribui diretamente para o incremento do lucro.
Ana necessita manter sua carga horária entre 26,67 e 53,33 horas para preservar a solução ótima. Dentro desse intervalo, as variações não produzem alterações significativas no resultado final.
Lídia tem um escopo de trabalho definido entre 15 e 30 horas.

# Questão 3
A melhor escolha para maximizar o lucro total é aumentar o número de horas de Lydia, sua váriavel dual é a maior entre os três, portanto, de uma hora trabalhada por ela terá um maior impacto na solução ótima.

# Questão 4

Sim, existe uma variável dual igual a zero. A restrição de horas trabalhadas de João é uma variável dual igual a zero. Este tipo de variável caracteriza que a restrição está não limitando a solução ótima. Portanto, aumentar ou reduzir o recurso associado a essa restrição não causaria impacto na solução ótima.